### PR TITLE
Multi-sweep joint refinement of unit cell params

### PR DESCRIPTION
**Pull request by benjaminhwilliams**
Add the option `multi_sweep_refinement`, which performs the same indexing as `multi_sweep_indexing` and additionally refines all sweeps together, restraining the unit cell parameters of each sweep to those of every other sweep.  This is achieved by setting the `dials.refine` option `refinement.parameterisation.crystal.unit_cell.restraints.tie_to_group.sigmas=.01,.01,.01,.01,.01,.01`, but other values and `tie_to_group`/`tie_to_target` schemes of `dials.refine` are exposed to xia2 via `xia2.Wrappers.Dials.Refine.Refine`.

Since this is likely to be most useful for multiple-sweep data in cases where each sweep contains few reflections, such as is typical in small-molecule chemical crystallography, the `multi_sweep_refinement` behaviour is made the default when `small_molecule=True`.

Detailed changes:
 * Introduce new optional PHIL parameter `multi_sweep_refinement` to xia2.  Make this `False` by default, except when `small_molecule=True`, in which case the default is `True`.  Ensure that if `multi_sweep_refinement` is enabled, `multiple_sweep_indexing` is also `True`, but retain the option to use the existing behaviour of multiple-sweep indexing without joint refinement.
 * Expose `dials.algorithms.refinement.restraints.restraints_parameterisation.uc_phil_scope` as part of the xia2 PHIL scope, to allow the user advanced control of the unit cell parameter restraints in multiple-sweep joint refinement.  Pass the restraint parameters to `dials.refine` via `xia2.Modules.Refiner.DialsRefiner.DialsRefiner.Refine` and `xia2.Wrappers.Dials.Refine.Refine`.
 * Allow a `xia2.Schema.XSample.XSample` to have a `multi_refiner` attribute, akin to a `multi_indexer`.  Ensure that the `multi_refiner` is serialisable.  If `multi_sweep_refinement=True`, make sure that if a new `xia2.Schema.XSweep.XSweep` is asked for its refiner and its associated `XSample` already has a `multi_refiner`, the `XSweep` returns the `multi_refiner`, rather than creating its own refiner afresh.
 * Allow a refiner to have multiple associated indexers.
 * Allow a refiner to have multiple associated integraters.
 * Ensure that a multiple-sweep `xia2.Schema.Interfaces.Refiner.Refiner` runs `dials.split_experiments` after refinement, in order to pass the individual refined experiment lists and reflection tables corresponding to each sweep to their respective integraters via the `_refinr_payload`.
 * Ensure that updating a refiner resets all associated integraters.

These changes are crudely illustrated by these flow schematics:

<p align="center">
DefaultMulti-sweep indexing
</p>

<p align="center">
Multi-sweep indexing and refinement
</p>

### CODE REVIEW

**Comment by benjaminhwilliams** [comment]
Sorry, I should have done a few squashes before creating this PR, to make the branch history a little more legible.  Naturally, this will be squash merged into master, so it doesn't matter in the longer term.

**Comment by benjaminhwilliams** [comment]
Hmm... I added an f-string.  Is this still not allowed?

**Comment by benjaminhwilliams** [comment]
I'm afraid I've done a lot of digging in xia2 trying to construct sensible tests and this is the best I could come up with.  If you want more unit-like tests, I'll need some hand holding.  Don't make me go back in there alone.

**Comment by Anthchirp** [comment]
xia2 master is now python 3.6+ only, so f-strings are very much allowed.

**Comment by graeme-winter** [comment]
Running off master on dials, with a good small molecule data set I see

```
Restraints will be ignored for scan_varying=True
There are 128 parameters to refine against 11947 reflections in 3 dimensions
Performing refinement of 4 Experiments...
```

in the second refinement run i.e. the scan_varying one (obviously 🙄) - is this what you would expect?

**Comment by graeme-winter** [comment]
> Running off master on dials, with a good small molecule data set I see
> 
> ```
> Restraints will be ignored for scan_varying=True
> There are 128 parameters to refine against 11947 reflections in 3 dimensions
> Performing refinement of 4 Experiments...
> ```
> 
> in the second refinement run i.e. the scan_varying one (obviously 🙄) - is this what you would expect?

Following discussion with @benjaminhwilliams  - @dagewa I gather that this message could be misleading since the unit cell restraints _are_ enforced? Should the message be revisited?

**Comment by benjaminhwilliams** [comment]
@dagewa, as I remember it from our discussion in January, scan-varying refinement disables all restraints _except_ the unit cell parameter restraints (controlled with the `refinement.parameterisation.crystal.unit_cell.restraints` PHIL parameters in `dials.refine`).  I seem to remember we had to fix a small bug for this.  I suspect the message is just a bit misleading.

**Comment by dagewa** [comment]
Unit cell restraints of the type you can set up with parameters in `restraints` PHIL scopes are only applied to static cells, not scan-varying. This is what the message means.

**Comment by dagewa** [comment]
Hmm, let me check. I think the scan-varying crystals don't get restraints, but restraints applied to static crystals persist. Maybe that was the fix, as before all restraints were nullified

**Comment by benjaminhwilliams** [comment]
I'm searching the commit history for context.

**Comment by benjaminhwilliams** [comment]
Ah, perhaps I missed something and should be specifying `crystal.unit_cell.force_static` in the `dials.refine scan_varying=True` call.

The commit that fixed a bug in `dials.refine` back in January was https://github.com/dials/dials/commit/0c7469d612a6ff6f8a1546fe2b960246eb043fb8.

**Comment by graeme-winter** [comment]
Following c560b136865f061efd99713b275f2bea8f1d28af I guess we should probably ensure that this is not performed by default? Since "general" data would benefit from scan varying refinement, only when you are dealing with sparse data would this help?

**Comment by benjaminhwilliams** [comment]
Thanks @graeme-winter, good spot.  Apologies @dagewa, I think too much water had gone under the bridge in the month between our discussion of https://github.com/dials/dials/commit/0c7469d612a6ff6f8a1546fe2b960246eb043fb8 and my eventual changes to `xia2.Wrappers.Dials.Refine.Refine` in aef10a230ee9d3b84533e685053f2a68047ab400.  Clearly it washed away some crucial synaptic links.

**Comment by dagewa** [comment]
Hi @benjaminhwilliams, no worries. I'm sort of surprised by this because I thought xia2 did the refinement steps static then scan-varying explicitly, so parameters could be set appropriately for each job. But maybe I am missing something too.

**Comment by benjaminhwilliams** [comment]
That's true indeed.  This new confusion is one of my making with this specific PR.  We are taking the chat offline tomorrow morning to get it straightened out.  Your input would be welcome if you're free.

File: Handlers/CommandLine.py [hunk 1]
```diff
@@ -267,22 +268,50 @@ def setup(self):
...
+                params.xia2.settings.small_molecule is True
+                and params.xia2.settings.indexer == "dials"
+            ):
+                PhilIndex.update("xia2.settings.multi_sweep_refinement=True")
+            else:
+                PhilIndex.update("xia2.settings.multi_sweep_refinement=False")
+            params = PhilIndex.get_python_object()
+
+        # Multi-sweep refinement requires multi-sweep indexing.
+        if params.xia2.settings.multi_sweep_refinement is True:
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Handlers/CommandLine.py:284]: You may be right, I think it went through a few wrangles at first, so originally one could get to this point and have `Auto` as a possible value.  That may no longer be the case, I'll double-check.

File: Handlers/CommandLine.py [hunk 2]
```diff
@@ -267,22 +268,50 @@ def setup(self):
...
             and params.xia2.settings.multiprocessing.mode == "parallel"
         ):
             logger.info(
-                "Multi sweep indexing disabled:\nMSI is not available for parallel processing."
+                "Multi sweep indexing disabled:\n"
+                "MSI is not available for parallel processing."
             )
             PhilIndex.update("xia2.settings.multi_sweep_indexing=False")
+            PhilIndex.update("xia2.settings.multi_sweep_refinement=False")
+            params = PhilIndex.get_python_object()
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Handlers/CommandLine.py:314]: Yes, I think that makes sense.  It's rather misleading to continue using a completely different strategy.  I'll make that change.

File: Handlers/Phil.py [hunk 3]
```diff
@@ -862,6 +863,13 @@
     .help = "Index all sweeps together rather than combining individual results " \
             "(requires dials indexer)"
     .expert_level = 2
+  multi_sweep_refinement = Auto
+    .type = bool
+    .help = "Refine all sweeps together, restraining unit cell parameters to their " \
+            "median, rather than refining image sequences individually (requires " \
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Handlers/Phil.py:869]: Sure thing.

File: Handlers/CommandLine.py [hunk 4]
```diff
@@ -267,22 +268,50 @@ def setup(self):
...
+                and params.xia2.settings.indexer == "dials"
+            ):
+                PhilIndex.update("xia2.settings.multi_sweep_refinement=True")
+            else:
+                PhilIndex.update("xia2.settings.multi_sweep_refinement=False")
+            params = PhilIndex.get_python_object()
+
+        # Multi-sweep refinement requires multi-sweep indexing.
+        if params.xia2.settings.multi_sweep_refinement is True:
+            PhilIndex.update("xia2.settings.multi_sweep_indexing=True")
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Handlers/CommandLine.py:285]: Hmm.  Good point.  OK.

File: Modules/Integrater/DialsIntegrater.py [hunk 5]
```diff
@@ -190,10 +190,24 @@ def _integrate_prepare(self):
 
         ## copy the data across
         refiner = self.get_integrater_refiner()
-        self._intgr_experiments_filename = refiner.get_refiner_payload("models.expt")
+        # For multi-sweep refinement, get the split experiments from after refinement.
+        if PhilIndex.params.xia2.settings.multi_sweep_refinement:
+            self._intgr_experiments_filename = refiner.get_refiner_payload(
+                "{}_models.expt".format(self._intgr_sweep._name)
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Integrater/DialsIntegrater.py:196]: Yes.  I'll do that.

File: Modules/Integrater/DialsIntegrater.py [hunk 6]
```diff
@@ -190,10 +190,24 @@ def _integrate_prepare(self):
...
+            )
+            self._intgr_indexed_filename = refiner.get_refiner_payload(
+                "{}_observations.refl".format(self._intgr_sweep._name)
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Integrater/DialsIntegrater.py:199]: Yep.

File: Modules/Refiner/DialsRefiner.py [hunk 7]
```diff
@@ -51,10 +55,55 @@ def Refine(self):
...
+
+        Add all the resulting experiment list/reflection table pairs to the payload.
+        """
+        split_experiments = SplitExperiments()
+        cwd = self.get_working_directory()
+        split_experiments.set_working_directory(cwd)
+        auto_logfiler(split_experiments, "SPLIT_EXPERIMENTS")
+        split_experiments.add_experiments(self._refinr_experiments_filename)
+        split_experiments.add_reflections(self._refinr_indexed_filename)
+        prefix = "{}_refined_split".format(split_experiments.get_xpid())
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Refiner/DialsRefiner.py:91]: Yep.

File: Modules/Refiner/DialsRefiner.py [hunk 8]
```diff
@@ -51,10 +55,55 @@ def Refine(self):
...
+        split_experiments._experiments_prefix = prefix
+        split_experiments._reflections_prefix = prefix
+        split_experiments.run()
+
+        num_sweeps = len(self._refinr_sweeps)
+        for i, sweep in enumerate(self._refinr_sweeps):
+            name = sweep._name
+            root = "{pre}_{ind:0{n_dig:d}d}".format(
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Refiner/DialsRefiner.py:99]: Yep.

File: Modules/Refiner/DialsRefiner.py [hunk 9]
```diff
@@ -51,10 +55,55 @@ def Refine(self):
...
+                pre=prefix, ind=i, n_dig=len(str(num_sweeps))
+            )
+            expts = os.path.join(cwd, root + ".expt")
+            refls = os.path.join(cwd, root + ".refl")
+            self.set_refiner_payload("{}_models.expt".format(name), expts)
+            self.set_refiner_payload("{}_observations.refl".format(name), refls)
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Refiner/DialsRefiner.py:105]: Yep.

File: Modules/Refiner/RefinerFactory.py [hunk 10]
```diff
@@ -27,14 +29,24 @@ def RefinerForXSweep(xsweep, json_file=None):
...
         assert os.path.isfile(json_file)
         logger.debug("Loading refiner from json: %s", json_file)
         refiner = refiner.__class__.from_json(filename=json_file)
 
-    refiner.add_refiner_sweep(xsweep)
+    if xsweep.sample.multi_refiner:
+        if not refiner._refinr_sweeps:
+            refiner._refinr_sweeps = xsweep.sample.get_sweeps()
+            for sweep in refiner._refinr_sweeps:
+                sweep._get_refiner()
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Refiner/RefinerFactory.py:47]: OK.

File: newsfragments/456.feature [hunk 11]
```diff
@@ -0,0 +1,7 @@
+Add the option ``multi_sweep_refinement`` to the DIALS pipelines.
+This performs the same indexing as ``multi_sweep_indexing`` and additionally refines all sweeps together, rather than refining each sweep individually.
+When refining the sweeps together, the unit cell parameters of each sweep are restrained to the mean unit cell.
+This is achieved by setting the ``dials.refine`` option ``refinement.parameterisation.crystal.unit_cell.restraints.tie_to_group.sigmas=0.01,0.01,0.01,0.01,0.01,0.01``, but other values and ``tie_to_group``/``tie_to_target`` schemes of ``dials.refine`` may be invoked by passing suitable parameters.
+See the various xia2 configuration parameters under ``dials.refine.restraints``, which are identical to the settings one can pass to ``dials.refine`` via its own parameter set ``refinement.parameterisation.crystal.unit_cell.restraints``.
+
+Since this is likely to be most useful for small-molecule chemical crystallography, the ``multi_sweep_refinement`` behaviour is made the default when ``small_molecule=True``.
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply newsfragments/456.feature:7]: Sure.

File: Modules/Refiner/DialsRefiner.py [hunk 12]
```diff
@@ -51,10 +55,55 @@ def Refine(self):
         refine.set_close_to_spindle_cutoff(
             PhilIndex.params.dials.close_to_spindle_cutoff
         )
+        # Do joint refinement of unit cell parameters if jointly refining multi sweeps.
+        # If the user hasn't specified any options, use the defaults from dials.refine.
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Refiner/DialsRefiner.py:59]: Ah you're right, there's some redundant working here.  I'll strip it back to the minimum.

**Comment by benjaminhwilliams** [comment]
> I'm not sure I fully understand the logic r.e. unit cell restraints - are these switched on or not by default when using multi_sweep_refinement=True? If yes, then is there a way to _not_ have them switched on?

Yes, at present the UC restraints are switched on by default when using `multi_sweep_refinement=True`.  This isn't achieved using a new parameter, but rather by passing non-null values to the parameter `tie_to_group.sigmas` of `dials.refine`.  I realise this means one cannot really turn the restraints off for multi-sweep refinement though...  Of course a cleaner and nicer solution would be to have a new parameter to switch this on and off.  I think this didn't occur to me at the outset because I didn't realise one might want to do joint refinement _without_ restraints.  In other words, the introduction of the `multi_sweep_refinement` parameter was sufficient to cover the only use case I envisaged and I didn't build in finer control.  I think we should regard this as a shortcoming of this PR and something to fix in future.  I suspect a separate issue may need to be raised.  (Actually, I suspect a lot of issues will need to be raised.)

> I note that the test for the new features only tests that the code runs, rather than testing that is behaves as expected. It might be worth extending the tests to cover the behaviour, although I appreciate this is can be tricky at times.

Yes, I tried quite hard to knock something into shape but my available time was quite fragmented and I made the mistake of trying to create unit tests, which the current architecture of xia2 makes difficult.  In the end, I gave up and resigned myself to these pretty unsatisfactory tests.  Quite honestly, I no longer have the strength to do anything cleverer solo.  I would need hand-holding.  I'm happy to do this as a pair-programming task at some point.

File: Handlers/CommandLine.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Handlers/CommandLine.py:284]: You're quite right, thanks for spotting this.

File: Modules/Refiner/RefinerFactory.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Refiner/RefinerFactory.py:47]: Hmm... Perhaps it's not necessary.  I think it may have crept in at a stage where I was getting very confused about when/why refiners were being created and reset.  It basically copies the logic at https://github.com/xia2/xia2/blob/ca2f126a668e2d0ffc7e5018f43b7a83b91c89e8/Modules/Indexer/IndexerFactory.py#L112-L115 Perhaps if this isn't necessary, nor is that?

Removing the line you refer to in `RefinerFactory` doesn't _seem_ to break anything...

[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Refiner/RefinerFactory.py:47]: Ah, yes it does.

[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Refiner/RefinerFactory.py:47]: I know this is a seemingly simple request but it is costing me blood sweat and tears.  I'm going to have to ignore it, sorry.

File: Modules/Refiner/DialsRefiner.py [hunk 13]
```diff
@@ -105,19 +156,22 @@ def _refine(self):
...
 
-            assert (
-                len(experiments.crystals()) == 1
-            )  # currently only handle one lattice/sweep
+            # currently only handle one lattice/refiner
+            assert len(experiments.crystals()) == 1
 
             scan_static = PhilIndex.params.dials.refine.scan_static
 
             # Avoid doing scan-varying refinement on narrow wedges.
```
[the remark this answers is not in the document]
Comment (benjaminhwilliams) [reply Modules/Refiner/DialsRefiner.py:162]: See comment in main thread.

**Comment by benjaminhwilliams** [comment]
> This pull request **introduces 1 alert** when merging [bcfa5d8](https://github.com/xia2/xia2/commit/bcfa5d89d07b4841a919b46528057bb7ffdf45a9) into [7ee6810](https://github.com/xia2/xia2/commit/7ee6810a5a18b9161423a9d548f6c32622bb3722) - [view on LGTM.com](https://lgtm.com/projects/g/xia2/xia2/rev/pr-c773704e6c869e8f7eca10e018a9582d4e373135)
> 
> **new alerts:**
> 
> * 1 for Unused local variable

This seems to be erroneous.  Perhaps LGTM cannot read into nested f-strings?

**Comment by benjaminhwilliams** [comment]
At present, it seems xia2 already does something quite sensible with regard to scan varying refinement of short scans.  If the scan is less than 36° (which is the default interval width for scan-varying refinement), then it sets the interval width to half the scan width, thereby ensuring that there are always at least two scan-varying refinement intervals in the scan.  In this PR, there is a tiny modification in that the interval width is set to half of the width of the _shortest_ scan.  This is forced on us by the fact that a single `dials.refine` call handles multiple sweeps, so we must accommodate the smallest sweep with whatever setting we choose.

A more sophisticated method would be to specify the number of scan intervals directly (and possibly choose one, not two, as the minimum).  `dials.refine` allows one to do this and it might be beneficial in cases where we expect too few reflections to be confident of doing a good job of refining the scan-varying unit cell parameters.  This might be especially useful for diamond anvil cell data, for example.  Assessing this, however, would require careful assessment and is out of scope for this PR.

In short, this PR does not alter the default behaviour of the scan-varying refinement step in xia2 and only affects the first, scan-static refinement step.
